pkg/ioutil/ioutil.go

```diff
@@ -64,6 +66,74 @@ func WriteOnClose(w io.Writer) *WriteOnCloser {
 	return &WriteOnCloser{w, false}
 }
 
+type ioret struct {
+	n   int
+	err error
+}
+
+// DeadlineWriter deadline writer with context
+type DeadlineWriter struct {
+	wc      WriterCloseWithError
+	timeout time.Duration
+	timer   *time.Timer
+	err     error
+}
+
+// WriterCloseWithError similar to io.WriteCloser
+// with additional CloseWithError implementation
+type WriterCloseWithError interface {
+	io.WriteCloser
+	CloseWithError(err error) error
+}
+
+// NewDeadlineWriter wraps a writer to make it respect given deadline
+// value per Write(). If there is a blocking write, the returned Writer
+// will return whenever the timer hits (the return values are n=0
+// and err=context.Canceled.)
+func NewDeadlineWriter(wc WriterCloseWithError, timeout time.Duration) WriterCloseWithError {
+	return &DeadlineWriter{wc: wc, timeout: timeout, timer: time.NewTimer(timeout)}
+}
+
+func (w *DeadlineWriter) Write(buf []byte) (int, error) {
+	if w.err != nil {
+		return 0, w.err
+	}
+
+	c := make(chan ioret, 1)
+
+	go func() {
+		n, err := w.wc.Write(buf)
+		c <- ioret{n, err}
+		close(c)
+	}()
+
+	select {
+	case r := <-c:
+		if !w.timer.Stop() {
+			<-w.timer.C
+		}
+		w.timer.Reset(w.timeout)
```

Comment:
I found this as per my testing. You actually need to reset this timer most preferably after go func() block above. Resetting the timer here means that next Write() call, select can pick the timer first.

Reply:
This is called only after the channel has been written to.. do you have a test case ?

Reply:
>  do you have a test case ?

yes, I have a fuse code which makes writing slow in one disk but upload will fail.

> This is called only after the channel has been written to

Correct, but what I am saying is that the timer could be expired already in the next Write() call so an error will be returned.

Reply:
@vadmeste but why worry about that since its a slow drive?